### PR TITLE
fix: remove the Sized requirement on ExecutionPlan::name()

### PR DESCRIPTION
**Pull request by waynexia**
## Which issue does this PR close?

Follow-up of https://github.com/apache/datafusion/pull/10266

## Rationale for this change

https://github.com/apache/datafusion/pull/10266 tries to provide a default implementation based on `static_name()`. This propagates the `Sized` requirement to `name()` and prohibits it from being called by a trait object.

## What changes are included in this PR?

Remove the default implementation of `ExecutionPlan::name()` and let every implementor implement their own based on `static_name()` to workaround object safety requirements. This is something like `as_any()`.

This solution is not elegant, but I can only find this way to suit both (1) invoke from trait object and (2) invoke from type name. Please add your suggestion if there is any better way to do so. Thanks in advance!

## Are there any user-facing changes?

Yes, for those who implements `ExecutionPlan`, they need to add a new method on `impl`

### CODE REVIEW

**Comment by matthewmturner** [comment]
Could we add a test to show that `name` can we called within a `ExecutionPlanVisitor` which i believe is a common use case.

File: datafusion/physical-plan/src/lib.rs [hunk 1]
```diff
@@ -116,12 +116,7 @@ pub mod udaf {
 /// [`required_input_ordering`]: ExecutionPlan::required_input_ordering
 pub trait ExecutionPlan: Debug + DisplayAs + Send + Sync {
     /// Short name for the ExecutionPlan, such as 'ParquetExec'.
-    fn name(&self) -> &'static str
-    where
-        Self: Sized,
-    {
-        Self::static_name()
-    }
+    fn name(&self) -> &'static str;
```
Comment (alamb) [inline datafusion/physical-plan/src/lib.rs:119]: If we are going to change this signature, perhaps we can change it to return a `&str` so anyone who implements ExeutionPlan can return whatever string they want. 

Like
```suggestion
    fn name(&self) -> &str;
```

File: datafusion/physical-plan/src/lib.rs [hunk 2]
```diff
@@ -116,12 +116,7 @@ pub mod udaf {
 /// [`required_input_ordering`]: ExecutionPlan::required_input_ordering
 pub trait ExecutionPlan: Debug + DisplayAs + Send + Sync {
     /// Short name for the ExecutionPlan, such as 'ParquetExec'.
```
Comment (alamb) [inline datafusion/physical-plan/src/lib.rs:118]: Perhaps we can also update the doc string to include a note about using `static_name` if desired to help people migrate

Comment (waynexia) [reply datafusion/physical-plan/src/lib.rs:118]: Great point 👍 I add an implementation note section in doc string

**Comment by waynexia** [comment]
> Could we add a test to show that `name` can we called within a `ExecutionPlanVisitor` which i believe is a common use case.

Certainly 👍  Add a compilation test `use_execution_plan_as_trait_object` in [a3bc903](https://github.com/apache/datafusion/pull/11047/commits/a3bc903ae72130ea0d18386b0dd616d9b76eaa60)

File: datafusion/physical-plan/src/lib.rs [hunk 3]
```diff
@@ -931,6 +940,14 @@ mod tests {
         assert_eq!(renamed_exec.name(), "MyRenamedEmptyExec");
         assert_eq!(RenamedEmptyExec::static_name(), "MyRenamedEmptyExec");
     }
+
+    /// A compilation test to ensure that the `ExecutionPlan::name()` method can
```
Comment (alamb) [inline datafusion/physical-plan/src/lib.rs:944]: 👍

**Review by alamb — APPROVED** [review]
Looks good to me -- thank you @waynexia  and @matthewmturner

**Review by alamb — APPROVED** [review]
Looks great to me -- thank you @waynexia  and @matthewmturner

**Comment by waynexia** [comment]
Thanks @matthewmturner @alamb :heart:

**Pull request merged**